Chapter 9 code - collision detection

louissalin opened the pull request.

louissalin committed:
add chapter 9 code

louissalin committed:
add ChopperSprite with spinning blades

louissalin committed:
add explosion particles

louissalin committed:
can deactivate emitters

louissalin committed:
fine tune sequence of steps when chopper explodes

louissalin committed:
change explosion texture and tweak emitter params

louissalin committed:
add bounding boxes and life loss events

louissalin committed:
can destroy choppers in game now

louissalin committed:
choppers flash when hit

louissalin committed:
generate 4 choppers that move on a path

louissalin committed:
quit game when player gets hit

louissalin committed:
make it easier to kill choppers. Add polish

louissalin committed:
restart game when player dies

louissalin committed:
make chopper a bit harder to kill

jcapellman commented on chapter-09/end/Engine/Particles/EmitterTypes/CircleEmitterType.cs:
```diff
@@ -0,0 +1,37 @@
+
+        public Vector2 GetParticlePosition(Vector2 emitterPosition)
+        {
+            var newAngle = _rnd.NextRandom(0, 2 * 3.14159f);
```
I assume you're going for PI here?  System.Math.PI is a handy constant for that

louissalin replied:
> Overall awesome job man!
> 
> Only common thing I noticed was the use of magic numbers instead of using the Texture2D dimensions.

I should probably tighten up the textures so they are the same size as the bounding boxes

jcapellman commented on chapter-09/end/Engine/Particles/Particle.cs:
```diff
@@ -0,0 +1,60 @@
+        // returns false if it went past its lifespan
+        public bool Update(GameTime gameTime)
+        {
+            // TODO: update rotation and scale
```
Later date item?

louissalin replied:
I should just remove this. It was there in chapter 8 too :)

jcapellman commented on chapter-09/end/Objects/BulletSprite.cs:
```diff
@@ -0,0 +1,22 @@
+        public BulletSprite(Texture2D texture)
+        {
+            _texture = texture;
+            AddBoundingBox(new Engine.Objects.BoundingBox(new Vector2(9, 4), 10, 22));
```
Is the texture these dimensions so we could just pull from the Texture2D object instead of the magic numbers?

louissalin replied:
no, the texture is bigger than the bullet itself, unfortunately. I hate those magic numbers though. There everywhere in this PR.

jcapellman commented on chapter-09/end/States/Dev/DevState.cs:
```diff
@@ -0,0 +1,72 @@
+                RemoveGameObject(_chopper);
+            }
+
+            if (_explosion != null && gameTime.TotalGameTime > TimeSpan.FromSeconds(10))
```
Braces here

jcapellman commented on chapter-09/end/States/Dev/DevState.cs:
```diff
@@ -0,0 +1,72 @@
+            if (_explosion != null && gameTime.TotalGameTime > TimeSpan.FromSeconds(10))
+                RemoveGameObject(_explosion);
+
+            if (_explosion != null)
```
Braces here too for consistency

jcapellman commented on chapter-09/start/chapter-08.sln:
```diff
@@ -0,0 +1,25 @@
+﻿
```
Probably should be named chapter-09.sln :)

louissalin replied:
haha, oops!

jcapellman commented on chapter-09/start/chapter-08.csproj:
```diff
@@ -0,0 +1,90 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Probably should be named chapter-09.csproj :)

jcapellman reviewed (changes requested):
Overall awesome job man!

Only common thing I noticed was the use of magic numbers instead of using the Texture2D dimensions.

jcapellman commented:
Will go through the code once more tomorrow night and test it myself - but again, awesome work!

louissalin committed:
refactor

remove some magic numbers by moving them to variables. Add a game object interface for object that cause damage. Extract a few functions in the gameplayState to make updateState function more readable.

louissalin committed:
improve chopper code

remove magic numbers and simplify flashing code

louissalin committed:
PR comments

louissalin committed:
redo chapter9 start dir

louissalin committed:
player should not be able to shoot when dead

jcapellman reviewed (approved):
Awesome work
